packages/protocol/test/liquidity/grandamento.ts

```diff
@@ -0,0 +1,1207 @@
+      })
+
+      it('stores the timestamp of the approval', async () => {
+        await grandaMento.approveExchangeProposal(proposalId, { from: approver })
+        const latestBlock = await web3.eth.getBlock('latest')
+        const proposal = parseExchangeProposal(await grandaMento.exchangeProposals(proposalId))
+        assertEqualBN(proposal.approvalTimestamp, latestBlock.timestamp)
+      })
+
+      it('does not remove the exchange proposal from the activeProposalIds linked list', async () => {
+        await grandaMento.approveExchangeProposal(proposalId, { from: approver })
+        assertEqualBNArray(await grandaMento.getActiveProposalIds(), [new BigNumber(1)])
+      })
+
+      it('reverts if the exchange proposal does not exist', async () => {
+        const nonexistentProposalId = 2
+        const proposal = parseExchangeProposal(
+          await grandaMento.exchangeProposals(nonexistentProposalId)
+        )
+        // As a sanity check, make sure the exchange is in the None state,
+        // indicating it doesn't exist.
+        assert.equal(proposal.state, ExchangeProposalState.None)
+        await assertRevertWithReason(
+          grandaMento.approveExchangeProposal(nonexistentProposalId, { from: approver }),
+          'Proposal must be in Proposed state'
+        )
+      })
+    })
+
+    it('reverts if called by anyone other than the approver', async () => {
+      await assertRevertWithReason(
+        grandaMento.approveExchangeProposal(proposalId, { from: accounts[2] }),
+        'Sender must be approver'
+      )
+    })
+  })
+
+  describe('#cancelExchangeProposal', () => {
+    const stableTokenSellAmount = unit.times(500)
+    describe('when called by the exchanger', () => {
+      beforeEach(async () => {
+        await grandaMento.createExchangeProposal(
+          stableTokenRegistryId,
+          stableTokenSellAmount,
+          false,
+          {
+            from: alice,
+          }
+        )
+      })
+
+      it('changes an exchange proposal from the Proposed state to the Cancelled state', async () => {
+        await grandaMento.cancelExchangeProposal(1, { from: alice })
+        const exchangeProposalAfter = parseExchangeProposal(await grandaMento.exchangeProposals(1))
+        assert.equal(exchangeProposalAfter.state, ExchangeProposalState.Cancelled)
+      })
+
+      it('reverts when the exchange proposal is not in the Proposed state', async () => {
+        // Get the exchange into the Approved state.
+        await grandaMento.approveExchangeProposal(1, { from: approver })
+        // Try to have Alice cancel it when the exchange proposal is in the Approved state.
+        await assertRevertWithReason(
+          grandaMento.cancelExchangeProposal(1, { from: alice }),
+          'Sender cannot cancel the exchange proposal'
+        )
+      })
+    })
+
+    describe('when called by the owner', () => {
+      beforeEach(async () => {
```

Comment:
nit: dedupe and move beforeEach up 1 describe

Reply:
Some of the specific cases need to do some things before creating an exchange proposal so keeping as is for now